### ISSUE TITLE
Report version of application and forms being used

### ISSUE THREAD

**Issue by diannakane** [medic/medic#1850]
Know what app and form version each user is using

Is there any way for LG to know which of the CHPs have updated their App? If not, can we make it such that if they do not update after a certain period, they can no longer proceed with the App until they update?

Clarification - do they mean synching data, or updating the forms, or updating the app itself?

Decide how managers and/or users find out that they have an out-of-date app. Decide what happens on the mobile device in this case.

Not for Feb, but start to consider user story

**Comment by diannakane** [comment]
@browndav do updates happen automatically when connected to internet or does the user need to do something to initiate the update? how long does an update typically take?

**Comment by ranjuts** [comment]
@browndav  any thoughts on the question from Dianna above? 

And, a related question --- would our webapp have the capability to detect the version of the CHW app when the CHP is connected to the internet and syncing data (or just whenever his/her phone is connected to the internet) Or would we need to figure out a manual way to do periodic checks on the app version each CHP is using?

**Comment by browndav** [comment]
We could insert version checks pretty much anywhere they're needed IIRC.

The application essentially consists of three parts: (i) a native Android container that needs to be updated via the Play Store, but contains a minimal amount of code and is forward-compatible (read: does not need to be updated to run newer versions of the web application); (ii) the actual web application, which uses the HTML 5 application cache (`appcache`) to automatically update itself whenever the application is closed and reopened; and (iii) replicated data, which includes some configuration and is intended to update continuously when internet connectivity is available.

For (ii), we also prompt the user to reload (read: update the application cache, and therefore most of the application) when internet connectivity is available; this works by detecting if the remote version is newer than the local version.

My instinct would be to block older versions only when we make a major breaking change, or when the versions are old enough (e.g. a few months) that we can't effectively test them anymore.

**Comment by ranjuts** [comment]
Thanks @browndav! Very helpful! Do you have any sense of how long it might take to reload/update?

**Comment by ranjuts** [comment]
Recommendation after design work: Include version numbers - browser and native android container-- in the analytics dashboard for LG Branch Managers. Fred mentioned that request has been made to add a “last active” widget. Alongside the widget, or in the same widget, if we could display the version number they are currently on (i.e. they were on when they were “last active”) that would be the best way for the branch managers to keep track of the app versions the CHPs are using. Asked @michaelkohn if this is technically feasible. His response below:

“If Dev is able to store the app version for that CHP in the database, then we should be able to use it in analytics.  My guess, and I may be totally wrong, is that it is not currently stored in the DB… so some dev time may be required to do so.  

When you update the issue, I would recommend confirming whether or not the version number is stored in the DB and can therefore make it’s way to PostgreSQL (where we get the data for analytics purposes).  We would also probably want to know what the most current version of the app is so we can highlight CHPs that aren’t on the the most current version (instead of just displaying the version they are on)” 

@abbyad @browndav could you confirm?
_______________________________________


Details captured during the investigation below: 

Do updates happen automatically when connected to internet or does the user need to do something to initiate the update? how long does an update typically take?
_Dave’s comment above in this thread._

Would our webapp have the capability to detect the version of the CHW app when the CHP is connected to the internet and syncing data (or just whenever his/her phone is connected to the internet)? Or would we need to figure out a manual way to do periodic checks on the app version each CHP is using?
_Dave’s comment above in this thread._

What is the difference between updating and reloading the app? -- both in terms of what it entails and what the CHP needs to do to complete each action.
_You’ll get an update/reload message - if you’re in the middle of a form, you’ll lose whatever data you typed in that form. Hit cancel on that message, go to the About page, and there’s reload button there._
_Feature request: in the about page show a note if a reload is needed to update the app._

As a CHP, if I cancel the update the first time, do I get another nudge the next time I open the app? If yes, how many times can I cancel before I stop getting these nudges?
_Next time you open the app, you will be updated whether you like it or not. Only valid reason to defer the update is if you’re in the middle of doing some work, in which case the correct behavior for the app is probably to update silently when you navigate away from the current form (alex and stefan)_

Do we know how long it usually takes (in the locality of our current users) to update and reload the app? 
_Incredibly quick to reload - the update message comes up after it already downloaded the update in the background. You just have to “restart” so to speak. 
It seems like CHPs are prompted when an update needs to be done. It takes practically no time to reload. The update is downloaded in the background, so there’s no “wait time” per se that might prevent CHPs from wanting to update._ 

What makes LG worried that some CHPs might not have the updated app if they are accessing the internet often enough to sync data in a timely manner. Is the actual worry that they might not be syncing the data frequently enough? If that’s the case, we have a different problem to solve. 
_According to Fred, this shouldn’t be an issue as the CHP have an incentive to be sync as often as they can. The branch managers review the CHPs performance dashboard at least twice a week, if the CHPs don’t sync as regularly they fall behind on their incentives, so it is highly unlikely that a CHP would be disconnected for too long. Also, they are at the branch office at least twice a week._

Knowing how often a CHP is interacting with the system. Is that sth the managers know already?
_A widget in analytics dashboard for branch managers will tell them when the CHP last synced newly completed forms._

Knowing if they’ve been online at all vs syncing data. Is it that they are not syncing data because there is nothing to sync or is it because for some reason they haven’t been able to sync because i) not been able to access data/internet ii)have accessed data/internet but some error in their hardware or app has prevented their data from syncing, or detecting internet/data connection. 
_Not really an issue in case of LG as the CHPs are at the branch office at least twice in a week to restock_

By “update”, do they mean syncing newly collected data (by the CHPs), updating the forms or updating the app itself? 
_Background on how this request came about from Fred: When we were doing the field test (December 2015), we saw a lot of this issue where when we were loading the app and it was shown as done, it wasn't happening consistently across all CHPs phones i.e. they were at different versions or even if the versions showed up as the same, the form content were different across the different phones. We followed the steps below to ensure everyone had the same version of the app and updated forms with the same content:_ 

_- Ensured that the URL is correct_ 
_- Ensured that versions for both the native android container and the browser are updated (there are two version numbers that you can see when you go to about page in the app-- this was implemented only after this issue was observed during the field test )_
_- Go to history, Tap “+” and count the number of forms to verify if all the updated forms were actually uploaded_

Do we have a sense of how often the CHPs are expected to and/or are able to sync the data, get connected to the internet? 
_Daily performance builds into their monthly performance; Branch managers look at the dashboard at least twice monthly; It’s in the CHPs best interest to sync their forms as often as possible; No predefined requirements/expectations as such._

After the app is completely stable (beyond the period when we are doing frequent iterations on the app itself), how often does LG expect to make changes to the forms? - perhaps they have yearly revisions or every six months etc? Apart from the forms, anything else that could be regularly revised/updated? If so how often?
_Not that we know of_

What would be the biggest concern for LG if all CHPs do not have the latest update of the app at all times? 
_That the data they are collecting and submitting don’t follow the updated format/structure_

How frequently do the CHPs travel to their respective branch offices? 

_- Monthly refresher meeting (happens at branch offices)_
_- As frequently as possible - refill of drugs (happens at the branch offices); very constant communication - could be upto daily_ 
_- At least twice in a week, that’s when they restock_

How often do they see their branch managers outside of their branch offices? 
_Depends on both on the branch manager and the CHP - if the branch manager feels like a certain CHP needs extra support and supervision then they tend to visit them more frequently than those who they feel are performing fine_

Related note but a different issue (re “last active” widget in analytics): Need to clearly define what last active means/should mean. Is it last connected/running the app vs last synced data? Also, perhaps another way to phrase it if it means last synced data. When Jane was in the field, CHPs complained how even though they were in the field working, using the app, visiting their clients, branch managers did not perceive them as “active” unless they synced their data. We cannot track all the times they are actually active as they are offline; we can at least communicate what we can track accurately.

**Comment by browndav** [comment]
The version and build number are stored in the database, but it applies to the server, not the mobile device. The mobile device's application version may not match that of the server's (e.g. if the device has been offline for 24 hours, and the server was updated during that time). Essentially, the information is trivial to get down to the mobile device (it's in the design document), but there isn't currently a nice *user-specific* place to return the information to server-side IIRC.

AFAICT, we'd need to modify the application to report each mobile device's view of the version number explicitly – likely at time-of-replication, since that's the only point where we can be sure that connectivity to the server is actually available. A separate request could fail due to the network dropping out during or immediately after replication, so IMO we'd have to create and maintain a local document containing the current application version, then include that as part of the usual filtered replication back up to the server. Even with this approach, I'd bet there are still timing issues between appcache and replication that could lead to an update not being reported.

Is there a lower-tech solution to this that's acceptable? Our timeline is already uncomfortably tight, and it seemed like we'd be coordinating updates with LG explicitly (as they'd presumably have training implications).

**Comment by SCdF** [comment]
Could we, on new ddoc, store the version in the user's db object? Do users
have write access to their own user object? This might not be too difficult

**Comment by sglangevin** [comment]
@browndav this issue isn't a requirement for the upcoming launch with LG, it's something they would like us to work on post-launch, which is why the design team has been investigating.

**Comment by abbyad** [comment]
Are we still looking at tracking version of each form and of app_settings?

**Comment by abbyad** [comment]
It seems as though we want to store the user's app versions (android container, webapp, ddoc) and last sync time in the user's doc in the db. This way it can be viewed in Analytics (easy to pull in) or eventually in the User profile or User config page (needs further design and dev).

This assumes that a user would have a single device - although that can perhaps be solved by storing these values along with a device id so we'd also know if a user was logging in with multiple devices.

**Comment by diannakane** [comment]
@alxndrsn is this feasible?

**Comment by alxndrsn** [comment]
>> It seems as though we want to store the user's app versions (android container, webapp, ddoc) and last sync time in the user's doc in the db

> is this feasible? 

Yes, all very simple except for last synch time, which won't have much meaning as we will only get the "last synch time" the next time synching happens!

**Comment by diannakane** [comment]
I'm not sure that I understand, Alex. Wouldn't the system take note of the
synch date/time when it's happening and that would remain as the last synch
time until a new synch date stamp replaces it?

On Fri, Apr 1, 2016 at 3:51 PM, Alex Anderson <notifications@github.com>
wrote:

> It seems as though we want to store the user's app versions (android
> container, webapp, ddoc) and last sync time in the user's doc in the db
>
> is this feasible?
>
> Yes, all very simple except for last synch time, which won't have much
> meaning as we will only get the "last synch time" the next time synching
> happens!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/medic/medic-webapp/issues/1850#issuecomment-204541626>
>



-- 
P.S. Join our community. Sign up for Medic Mobile's newsletter
<http://medicmobile.us8.list-manage.com/track/click?u=e5b07d2afb4b273eaff8e3c53&id=a3129fc799&e=c511575255>
.


Dianna Kane | Senior Designer
Medic Mobile <http://www.medicmobile.org/> +1 216 262 2204
@*M*edic <http://www.twitter.com/medic> Facebook
<https://www.facebook.com/MedicMobile> @diannakane

**Comment by garethbowen** [comment]
@SCdF Is it just me, or would this be _really_ easy to solve by adding the app version and form revs to the telemetry metadata?

**Comment by SCdF** [comment]
@garethbowen it astounds me that we didn't at least already have the app version in the mix! How did we forget to do that!?

I've added the app version.

For xml forms I've used the md5 digest hash: `_rev` doesn't work for us because PouchDB just makes them up, and the versions inside the XML itself are neither required nor regulated (most forms don't use them, and the ones that do don't enforce it changing if the content changes and vice versa).

There is no clear way of versioning JSON forms: we don't have settings versions that are cross-device. However, we could whiteboard some jankier ideas (hash a `JSON.stringify` per form in the settings), or deal with it at a higher level (automatically version settings as part of the medic-conf flow would be my vote).

https://github.com/medic/medic/pull/5494

**Comment by SCdF** [comment]
(I haven't moved this ticket or assigned it etc because I'm not sure this covers the whole deal: but it's definitely related and definitely useful.

I'm happy to create a new ticket for just this PR if you'd prefer)

**Comment by garethbowen** [comment]
There is a lot of information in this issue. For now we're going to implement a simplified solution which captures the version information and stores it in CouchDB and resolve this issue. If you'd like to integrate with dashboards or anywhere else please raise a new issue and we can do it in a later version.

**Comment by garethbowen** [comment]
Ready for AT in the `telemetry_app_version` branch.

Make sure when the telemetry is sent at the end of the month you can see the local app version and all the _revs of the forms.

**Comment by ngaruko** [comment]
LGTM @
_**before**_
![image](https://user-images.githubusercontent.com/6979995/55929271-0155bf00-5c70-11e9-95a6-f2e94a1cc437.png)

**_after_**

![image](https://user-images.githubusercontent.com/6979995/55933302-01aa8600-5c81-11e9-8cd7-6b1613bbed7e.png)

**Comment by SCdF** [comment]
Merged.
